nucypher/blockchain/eth/clients.py

```diff
@@ -15,16 +16,109 @@
 )
 from geth.process import BaseGethProcess
 from twisted.logger import Logger
+from web3 import Web3
 
-from nucypher.config.constants import DEFAULT_CONFIG_ROOT, BASE_DIR, DEPLOY_DIR, USER_LOG_DIR
+from nucypher.config.constants import DEFAULT_CONFIG_ROOT, DEPLOY_DIR, USER_LOG_DIR
 
 NUCYPHER_CHAIN_IDS = {
     'devnet': 112358,
 }
 
 
-class NuCypherGethProcess(LoggingMixin, BaseGethProcess):
+class Web3ClientError(Exception):
+    pass
+
+
+class Web3ClientConnectionFailed(Web3ClientError):
+    pass
+
+
+class Web3ClientUnexpectedVersionString(Web3ClientError):
+    pass
+
+
+class Web3Client(object):
+
+    def __new__(self, w3: Web3, *args):
+        #
+        # *Client version format*
+        # Geth Example: "'Geth/v1.4.11-stable-fed692f6/darwin/go1.7'"
+        # Parity Example: "Parity//v1.5.0-unstable-9db3f38-20170103/x86_64-linux-gnu/rustc1.14.0"
+        # Ganache Example: "EthereumJS TestRPC/v2.1.5/ethereum-js"
+        #
+        client_data = w3.clientVersion.split('/')
+        node_technology = client_data[0]
+
+        GETH = 'Geth'
+        PARITY = 'Parity'
+        GANACHE = 'EthereumJS TestRPC'
+
+        try:
+            gclass = {
+                GETH: GethClient,
+                PARITY: ParityClient,
+                GANACHE: GanacheClient,
+            }[node_technology]
+            return gclass(w3, *client_data)
+
+        except KeyError:
+            raise NotImplementedError(node_technology)
+
+
+class Web3ClientBase(object):
+
+    def __init__(self, w3, node_technology, version, backend, **kwargs):
```

Comment:
`client_data` can sometimes split into a list of size 4 not 3 - eg. for Geth and Parity client version strings; does that cause a problem here? This was the reason for the tricky splitting logic dependent on node_technology.

Reply:
Yes I left this for when we have a way to test the various providers

Reply:
Ok - but wouldn't this currently produce an error due to the number of parameters provided based on Geth and Parity `clientVersion` strings?

Reply:
Yeah - We'll need to get started on a corresponding mock web3 clients to see it in action.  Some of these init params can be optional with a constant default if needed